src/rulesets/Base/fastmath_able.jl

```diff
@@ -36,10 +36,94 @@ let
 
 
         # Unary complex functions
-        @scalar_rule abs(x::Real) sign(x)
-        @scalar_rule abs2(x) 2x
-        @scalar_rule angle(x::Real) Zero()
-        @scalar_rule conj(x::Real) One()
+        ## abs
+        function frule((_, Δx), ::typeof(abs), x::Real)
+            return abs(x), sign(x) * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs), z::Complex)
+            Ω = abs(z)
+            return Ω, (real(z) * real(Δz) + imag(z) * imag(Δz)) / ifelse(iszero(z), one(Ω), Ω)
+        end
+        
+        function rrule(::typeof(abs), x::Real)
+            function abs_pullback(ΔΩ)
+                return (NO_FIELDS, real(ΔΩ)*sign(x))
+            end
+            return abs(x), abs_pullback
+        end
+        function rrule(::typeof(abs), z::Complex)
+            Ω = abs(z)
+            function abs_pullback(ΔΩ)
+                Δu = real(ΔΩ)
+                return (NO_FIELDS, Δu*z/ifelse(iszero(z), one(Ω), Ω))
+            end
+            return Ω, abs_pullback
+        end
+
+        ## abs2
+        function frule((_, Δx), ::typeof(abs2), x::Real)
+            return abs2(x), 2x * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs2), z::Complex)
+            return abs2(z), 2 * (real(z) * real(Δz) + imag(z) * imag(Δz))
+        end
+        
+        function rrule(::typeof(abs2), x::Real)
+            function abs2_pullback(ΔΩ)
+                return (NO_FIELDS, 2real(ΔΩ)*x)
+            end
+            return abs2(x), abs2_pullback
+        end
+        function rrule(::typeof(abs2), z::Complex)
+            function abs2_pullback(ΔΩ)
+                Δu = real(ΔΩ)
+                return (NO_FIELDS, 2real(ΔΩ)*z)
+            end
+            return abs2(z), abs2_pullback
+        end
```

Comment:
As far as I can tell, this one is exactly the same as the `Real` version, and line 79 is unnecessary.

Reply:
Good catch!